### PR TITLE
Issue 22574 - Compiler segfaults on `template test(x* x)`

### PR DESCRIPTION
**Pull request by maxhaton**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]
If this fixes issue 22574 then please make the commit message say "Fix issue 22574: ..."

File: test/fail_compilation/test22574.d [hunk 1]
```diff
@@ -0,0 +1,13 @@
+//https://issues.dlang.org/show_bug.cgi?id=22574
+/*
+TEST_OUTPUT:
+---
+fail_compilation/test22574.d(100): Error: variable `x` is used as a type
+fail_compilation/test22574.d(100):        variable `x` is declared here
```
Comment (RazvanN7) [inline test/fail_compilation/test22574.d:6]: Can we agree that this is a weird error message? When looking at `x* x` the compiler should first try to analyze the type and then the variable name, however, in this case it seems that the name is inserted in the symbol table before actually resolving the type. I would expect that the error message is something along the lines of "type `x` is not defined".

Comment (RazvanN7) [reply test/fail_compilation/test22574.d:6]: Looking in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2560) you can see that when the parameter is [declared](https://github.com/dlang/dmd/blob/master/src/dmd/dtemplate.d#L5534) (for template value parameters) the identifier is inserted in the scope before actually doing any semantic on the type. However, the next thing that it is done in [template declaration semantic](https://github.com/dlang/dmd/blob/master/src/dmd/dsymbolsem.d#L2566) is to do [semantic on the type](https://github.com/dlang/dmd/blob/master/src/dmd/templateparamsem.d#L78). So, if we want to fix the underlying issue and report a correct error, we should add

```d
if (valtype)
    valtype = valtype.typeSemantic(loc, sc);
```

in TemplateValueParameter::declareParameter.

That will make the segfault go away and output: "Undefined identifier `x`"

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
See my previous comments

File: test/fail_compilation/test22574.d (hunk 1 above)
Comment (maxhaton) [reply test/fail_compilation/test22574.d:6]: This is why I left the comment in the source code.

**Review by RazvanN7 — APPROVED** [review]
Thanks!

**Comment by maxhaton** [comment]
@RazvanN7 I have changed it as you suggested, however this has led to some changes in error messages in other (failing, to be clear) template value parameter tests. I have corrected these error messages since I think they're basically reasonable still

**Comment by RazvanN7** [comment]
Please add a commit message that contains "Fix Issue 22574" so that the bot picks it up.

**Pull request merged**